enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+a IAM role the pod should assume. When a pod runs with the annoated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contanis the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+3. OpenShift should create the OIDC provider in IAM
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seemsless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
+   machine-api) to use bound tokens
+
+## Proposal
+
+A new `aws-pod-identity-webhook` image will be added to the release payload
+based on [this](https://github.com/openshift/aws-pod-identity-webhook) repo.
+
+Considerable functionality will be added to the `cloud-credential-operator`
+(CCO).  This functionality is AWS-only and will only be active for clusters
+deployed with `platform: aws`.
+
+- Management of the `aws-pod-identity-webhook` and its dependent resources
+  - Synchronization of the `spec.clientConfig.caBundle` on
+    `MutationWebhookConfiguration` whenever the service serving CA
+    (`service-ca`) rotates
+- Creation of the OIDC provider in IAM (pending on Open Question)
+- Synchronization of the `keys.json` in the S3 bucket pointed to by the OIDC
+  provider with the rotatable public keys contained in
+  `bound-sa-token-signing-certs` configmap in the `openshift-kube-apiserver`
+  namespace (pending on Open Question)
+
+IAM permissions will need to be expanded to include IAM OIDC provider create and
+S3 bucket create/write if the CCO is to manage the OIDC provider (see Open
+Questions section)
+
+In terms of API changes, an additional `serviceAccountRef` field will be added
+to the `CredentialsRequests` CR `spec` alongside the existing [`secretRef`
+field](https://github.com/openshift/cloud-credential-operator/blob/9cc1c1abf898cfa08429d465da0f02d064bd89ee/manifests/00_v1_crd.yaml#L37-L42).
+Usage of this field is an indication by the user that they wish to use this new
+bound token injection mechanism.  The CCO will create a role, instead of a user,
+in response and create a `ServiceAccount` with the appropriate role annotation.
+A pod started with this `ServiceAccount` will have a token injected by the
```

Comment:
what if the pod is running and waiting on service account before the mutation webhook is even running?

Reply:
we will need to gate the processing of CredentialRequests CRs that request this injection method on the readiness of the webhook i.e. don't create the serviceaccount until the webhook is ready.

Reply:
On second though, because the webhook only processes the pod at creation time, even though they won't run until their serviceaccount is created, they can get past admission before the webhook is ready. Hmm... I need to see how the system behaves when you try to start a pod whose serviceaccount doesn't exist.